drivers/usb/udc/udc_dwc2.c

```diff
@@ -2017,6 +2233,14 @@ static const struct udc_api udc_dwc2_api = {
 		.ghwcfg4 = DT_INST_PROP(n, ghwcfg4),				\
 	};									\
 										\
+	IF_ENABLED(CONFIG_DCACHE, (BUILD_ASSERT(				\
+		!IS_ENABLED(CONFIG_UDC_DWC2_DMA) ||				\
+		(DT_INST_PROP(n, ghwcfg2) & USB_DWC2_GHWCFG2_OTGARCH_MASK) !=	\
+			(USB_DWC2_GHWCFG2_OTGARCH_INTERNALDMA <<		\
+			 USB_DWC2_GHWCFG2_OTGARCH_POS),				\
+		"Data Cache management is not supported yet");			\
+	))									\
+										\
```

Comment:
What does it do?

Reply:
Ensures that `CONFIG_UDC_DWC2_DMA` will fail compilation if `CONFIG_DCACHE` is enabled and DWC2 implementation supports Internal DMA.

It is important do fail here, because USB device stack does not support alignment and padding to data cache line. When the USB stack will have proper D-Cache aware mechanism in place, this build-time assert can be removed.

Reply:
That is not obvious. Also you are using `sys_cache_data_.*` but BUILD_ASSERTs here, IMO that does not make sense. I suggest to use preprocessor warning and not BUILD_ASSERTs.

Reply:
The `sys_cache_data_` does not help if the buffers are not cache line aligned and padded. The change necessary is outside the driver. It is known that the driver will fail when D-Cache is enabled and the known issues are best to be communicated by failed build.

Reply:
Changed the error from "Data Cache management is not supported yet" to "USB stack does not provide D-Cache aware buffers yet" to better reflect the reason behind it.